synapse/handlers/federation.py

```diff
@@ -171,6 +171,23 @@ def on_receive_pdu(self, origin, pdu, get_missing=True):
                         yield self._get_missing_events_for_pdu(
                             origin, pdu, prevs, min_depth
                         )
+
+                        # Update the set of things we've seen after trying to
+                        # fetch the missing stuff
+                        have_seen = yield self.store.have_events(
+                            [ev for ev, _ in pdu.prev_events]
+                        )
+
+                        seen = set(have_seen.keys())
```

Comment:
haven't we already got all this at https://github.com/matrix-org/synapse/pull/2170/files#diff-3f6de39b112e0c501c3c2925a571b908R199 ?